### PR TITLE
WIP Firebase

### PR DESCRIPTION
**Pull request by mjvalade**
@brittanystoroz

A user can create a new hiking trail in the app that shows up in the redux store but need to send it to the DB as well.

Trying to adapt existing createTrail action and trails-reducer files from redux-firebase repo expenses files to connect them to firebase.

### CODE REVIEW

**Comment by brittanydionigi** [comment]
Unrelated issue: make sure you add `redux` and `react-redux` to your package.json. It looks like you probably installed them but didn't add the little `--save` flag to persist them to your package.json. I do that all the time :)

**Comment by brittanydionigi** [comment]
Second issue: You'll want to install and use `redux-thunk` as middleware in your redux store. This will allow you to add a redux action that can perform the firebase functionality before handing your trail data off to the reducer.

**Comment by brittanydionigi** [comment]
In your `/lib/actions/index.js`, you'll want to import firebase and get a reference to the database table where you want to store your trail data:

``` javascript
import firebase from 'firebase';

const firebaseTrails = firebase.database().ref('trails');
```

It's ok if `trails` doesn't exist in your database yet -- if Firebase can't find it, it will automatically create that key for you. Once you've successfully saved a trail to the database, you'll be able to open the database page of your firebase project and you'll see the new trail entered under a key called "trails".

**Comment by brittanydionigi** [comment]
Using redux-thunk, you can modify your createTrail function to look more like this:

``` javascript
export const createTrail(trailData) {
  return (dispatch) => {

    // we need to create a new unique key to store our trail
    // data under. this will tell firebase to put this particular
    // trail that we're saving under trails/<unique-key>
    let newTrailKey = firebaseTrails.push().key;

     // access the new "table row" we created under `trails/<unique-key`
     // and set the data for it
      firebaseTrails.child(newTrailKey).set(trailData)
      .then(() => {  // when the write to the database was successful, finally dispatch our action
        dispatch({
          type: 'CREATE_NEW_TRAIL',
          trail: trailData
        });
      })
      .catch(error => {
        console.log("Error saving trail: ", error);
      });
  }
};
```

**Comment by mjvalade** [comment]
@martensonbj and @stevekinney

Sorry to keep harping on this redux but my store is empty. I can make a trail and send it to firebase but for some reason it seems to bypass the store so I can't then pull anything out to render on the page. I'm referencing the weather project, Brittany's example repo, and reading docs. Still not sure why it's happening (or not happening). Thanks.

**Comment by stevekinney** [comment]
Just for clarification: is it not making it into the store or is it just not rendering on the page? A quick way to check is to see if it's in the store in the Redux debugging tools for Chrome.

**Comment by mjvalade** [comment]
When I check the Redux tools in Chrome the trails array in the store is empty. But there are three trails in firebase.

**Comment by brittanydionigi** [comment]
Can you verify that your action is hitting your reducer? Can you verify what data is being sent when you dispatch the `CREATE_NEW_TRAIL` action?

File: lib/actions/index.js [hunk 1]
```diff
@@ -1,6 +1,38 @@
...
+export const createTrail = (trailData) => {
+  return (dispatch) => {
+
+    let newTrailKey = firebaseTrails.push().key;
+
+      firebaseTrails.child(newTrailKey).set(trailData)
+      .then(() => {
+        dispatch({
+          type: 'CREATE_NEW_TRAIL',
+          trail: trailData
```
Comment (brittanydionigi) [inline lib/actions/index.js:15]: Verify that `trailData` is an object that contains the data you want sent to your reducer. Even if it's just adding a `console.log` before the `dispatch` call.

File: lib/store/reducers/trails-reducer.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import { initialState } from '../initialState';
+
+export default function trailsReducer(state = initialState.trails, action) {
+  switch(action.type) {
+    case 'RECEIVE_ALL_TRAILS':
+      return {
+        data: action.trails
+      };
+
+    case 'CREATE_NEW_TRAIL':
```
Comment (brittanydionigi) [inline lib/store/reducers/trails-reducer.js:8]: Can you verify that this is being hit when you create a new trail?

**Comment by brittanydionigi** [comment]
This pull request is really difficult to read because there is so much else going on in it.  If the code changes don't relate to the action & reducer that are giving you problems, they shouldn't be in this pull request. e.g. any commented out code should not be in this PR and any of your HTML/component/CSS changes shouldn't be in this pull request. Fine for now, but in the future, make sure any code changes going into a PR are specific to the problem that you are having. Either stash the rest of the changes, or commit them to another branch to save for later.

(i.e. you should have merged this pull request after the firebase problems were sorted out, and then opened a new one with your redux store problems)

**Pull request merged**